src/doc/trpl/the-stack-and-the-heap.md

```diff
@@ -1,3 +1,526 @@
+| Address | Name | Value  |
++---------+------+--------+
+| 2 ^ 30  |      | 5      |
+| ...     | ...  | ...    |
+| 1       | y    | 42     |
+| 0       | x    | 2 ^ 30 |
+
+We have `2 ^ 30` in our hypothetical computer with 1GB of RAM. And since our
+stack grows from zero, the easiest place to allocate memory is from the other
+end. So our first value is at the highest place in memory. And the value of the
+struct at `x` has a [raw pointer][rawpointer] to the place we’ve allocated on
+the heap, so the value of `x` is `2 ^ 30`, the memory location we’ve asked for.
+
+[rawpointer]: raw-pointers.html
+
+Since this memory is on the heap, it can stay alive longer than the function
+which allocates the box. In this case, however, it doesn’t. When the function
+is over, we need to free the stack frame for `main()`. `Box<T>`, though, has a
+trick up its sleve: [Drop][drop]. The implementation of `Drop` for `Box`
+deallocates the memory that was allocated when it was created. Great! So when
+`x` goes away, it first frees the memory allocated on the heap:
+
+[drop]: drop.html
+
+| Address | Name | Value  |
++---------+------+--------+
+| 1       | y    | 42     |
+| 0       | x    | ?????? |
+
+And then the stack frame goes away, freeing all of our memory.
+
+# Arguments and borrowing
+
+We’ve got some basic examples with the stack and the heap going, but what about
+function arguments and borrowing? Here’s a small Rust program:
+
+```rust
+fn foo(i: &i32) {
+    let z = 42;
+}
+
+fn main() {
+    let x = 5;
+    let y = &x;
+
+    foo(y);
+}
+```
+
+When we enter `main()`, memory looks like this:
+
+| Address | Name | Value |
++---------+------+-------+
+| 1       | y    | 0     |
+| 0       | x    | 5     |
+
+`x` is a plain old `5`, and `y` is a reference to `x`. So its value is the
+memory location that `x` lives at, which in this case is `0`.
+
+What about when we call `foo()`, passing `y` as an argument?
+
+| Address | Name | Value |
++---------+------+-------+
+| 3       | z    | 42    |
+| 2       | i    | 0     |
+| 1       | y    | 0     |
+| 0       | x    | 5     |
+
+Stack frames aren’t just for local bindings, they’re for arguments too. So in
+this case, we need to have both `i`, our argument, and `z`, our local variable
+binding. `i` is a copy of the argument, `y`. Since `y`’s value is `0`, so is
+`i`’s.
+
+This is one reason why borrowing a variable doesn’t deallocate any memory: the
+value of a reference is just a pointer to a memory location. If we got rid of
+the underlying memory, things wouldn’t work very well.
+
+# A complex example
+
+Okay, let’s go through this complex program step-by-step:
+
+```rust
+fn foo(x: &i32) {
+    let y = 10;
+    let z = &y;
+
+    baz(z);
+    bar(x, z);
+}
+
+fn bar(a: &i32, b: &i32) {
+    let c = 5;
+    let d = Box::new(5);
+    let e = &d;
+
+    baz(e);
+}
+
+fn baz(f: &i32) {
+    let g = 100;
+}
+
+fn main() {
+    let h = 3;
+    let i = Box::new(20);
+    let j = &h;
+
+    foo(j);
+}
+```
```

Comment:
While I am *somewhat* hesitant to suggest making a complex example even more complex, some small changes to this program would give you an opportunity to directly illustrate how stack-allocation and heap-allocation differ.

In particular, an example where a box is freed and then *its* memory is reused for another box ... I think that would crystallize the most important points here, especially if the freeing and reuse is done in a *non-LIFO* fashion.

I am thinking that you could revise the example this way ([playpen](http://is.gd/mzGe3q)):
```rust
fn foo(v: &i32, w: lb::LoudBox<i32>) {
    let x = 10;
    let y = &x;
    let z = lb::LoudBox::new(40);
    baz(y, w);
    bar(v, y);
}

fn bar(a: &i32, b: &i32) {
    let c = 5;
    let d = lb::LoudBox::new(5);
    let e = &d.0;
    let f = lb::LoudBox::new(6);

    baz(e, f);
}

fn baz(g: &i32, h: lb::LoudBox<i32>) {
    let i = 100;
}

fn main() {
    let j = 3;
    let k = lb::LoudBox::new(20);
    let l = lb::LoudBox::new(30);
    let m = &j;

    foo(m, l);
}

/// This module just instruments what addresses are taken and returned
/// from an abstract pool of word-oriented addresses.
mod lb {
    // This models a bitmap that one might actually see in some real-world allocators.
    // (In our case, this tiny address space has only 32 entries available.)
    static mut TAKEN: i32 = 0;

    pub struct LoudBox<T: ::std::fmt::Debug>(pub Box<T>, i8);

    impl<T: ::std::fmt::Debug> LoudBox<T> {
        pub fn new(t: T) -> LoudBox<T> {
            let addr;
            'find_addr: loop {
                for i in 0..32 {
                    unsafe {
                        if 0 == (TAKEN & 1 << i) {
                            addr = i;
                            TAKEN = TAKEN | 1 << i;
                            break 'find_addr;
                        }
                    }
                }
                panic!("tiny address space exhausted");
            }
            println!("Creating Box({:?}) at {}", t, addr);
            LoudBox(Box::new(t), addr)
        }
    }
    
    impl<T: ::std::fmt::Debug> Drop for LoudBox<T> {
        fn drop(&mut self) {
            println!("Dropping Box({:?}) at {}", self.0, self.1);
            unsafe { TAKEN = TAKEN & !(1 << self.1); }
        }
    }
}
```
The crucial change, illustrated by the output of the instrumentation, is that we get to *reuse* the address for the boxed memory at address 1, even though there are boxes at memory locations 0 and 2 that are still in use; thus, we see that heap-memory does not follow the LIFO ordering.

(The main drawback to this change is if you start moving boxes around, then you're going to have to at least mention that when a `Box<_>` is moved, the generated code somehow knows that it has been moved.  So maybe you could put this new example in a new section, after this current one.)